src/uuid.cr

```diff
@@ -176,6 +206,119 @@ struct UUID
+  #   monotonically increasing, with only 14 bits of the clock sequence being used effectively.
+  #   The clock sequence should be stored in a stable location, such as a file. If it is not
+  #   stored, a random value is used by default. If not provided the current time milliseconds
+  #   are used.
+  # * `node_id`: In case the traditional MAC address based approach should be taken the
+  #   `node_id` can be provided. Otherwise secure random is used.
+  def self.v1(*, clock_seq : UInt16? = nil, node_id : MAC? = nil) : self
+    tl = Time.local
+    now = (tl.to_unix_ns / 100).to_u64 + 122192928000000000
+    seq = ((clock_seq || (tl.nanosecond/1000000).to_u16) & 0x3fff) | 0x8000
+
+    time_low = UInt32.new(now & 0xffffffff)
+    time_mid = UInt16.new((now >> 32) & 0xffff)
+    time_hi = UInt16.new((now >> 48) & 0x0fff)
+    time_hi |= 0x1000 # Version 1
+
+    uuid = uninitialized UInt8[16]
+    IO::ByteFormat::BigEndian.encode(time_low, uuid.to_slice[0..3])
+    IO::ByteFormat::BigEndian.encode(time_mid, uuid.to_slice[4..5])
+    IO::ByteFormat::BigEndian.encode(time_hi, uuid.to_slice[6..7])
+    IO::ByteFormat::BigEndian.encode(seq, uuid.to_slice[8..9])
+
+    if node_id
+      6.times do |i|
+        uuid.to_slice[10 + i] = node_id[i]
+      end
+    else
+      Random::Secure.random_bytes(uuid.to_slice[10..15])
+      # set multicast bit as recommended per section 4.5 of the RFC 4122 spec
+      # to not conflict with real MAC addresses
+      uuid[10] |= 0x01_u8
+    end
+
+    new(uuid, version: UUID::Version::V1, variant: UUID::Variant::RFC4122)
+  end
+
+  # Generates RFC 4122 v2 UUID.
+  #
+  # Version 2 UUIDs are generated using the current time, the local machine’s MAC address,
+  # and the local user or group ID. However, they are not widely used due to their limitations.
+  # For a given domain/id pair, the same token may be returned for a duration of up to 7 minutes
+  # and 10 seconds.
+  #
+  # * `domain`: The `Domain` of the uuid.
+  # * `id`: The id depends on the domain, for the `Domain::Person` usually the local user id
+  #    is used, for `Domain::Group` usually the local group id is used.
+  # * `node_id`: In case the traditional MAC address based approach should be taken the
+  #   `node_id` can be provided. Otherwise secure random is used.
+  def self.v2(domain : Domain, id : UInt32, node_id : MAC? = nil) : self
+    uuid = v1(node_id: node_id).bytes
+    uuid[6] = (uuid[6] & 0x0f) | 0x20 # Version 2
+    uuid[9] = domain.to_u8
+    IO::ByteFormat::BigEndian.encode(id, uuid.to_slice[0..3])
+    new(uuid, version: UUID::Version::V2, variant: UUID::Variant::RFC4122)
+  end
+
+  # Generates RFC 4122 v3 UUID.
+  #
+  # * `name`: The name used to generate the UUID, it can be a string of any size.
+  # * `namespace`: Specifies the type of the name, usually one of `Namespace`
+  def self.v3(name : String, namespace : UUID) : self
+    hash = Digest::MD5.digest do |ctx|
```

Comment:
This introduces a dependency on OpenSSL, which `Digest::*` relies on. Instead it should use the private `Crystal::Digest::MD5` with `require "crystal/digest/md5"`. Alternatively you could follow what `HTTP::WebSocket::Protocol` does:

https://github.com/crystal-lang/crystal/blob/f6266db87dd6fff89e3e9af7e31b0eb22a7c7451/src/http/web_socket/protocol.cr#L5-L9

https://github.com/crystal-lang/crystal/blob/f6266db87dd6fff89e3e9af7e31b0eb22a7c7451/src/http/web_socket/protocol.cr#L356-L360

Reply:
My first attempt failed, wasm still red, seems as if `::Crystal::Digest::SHA1` uses `Digest` which does require `openssl`, will continue until it is resolved.

Reply:
Seems my initial assumption was wrong the issue is related to the  `flag?(:without_openssl)` seems as if this is not working for the wasm32 while using `flag?(:wasm32)` does work.